compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUselessTypeOperationCallChecker.kt

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.expression
+
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.expressions.*
+import org.jetbrains.kotlin.fir.resolve.fullyExpandedType
+import org.jetbrains.kotlin.fir.resolve.inference.isFunctionalType
+import org.jetbrains.kotlin.fir.typeContext
+import org.jetbrains.kotlin.fir.types.*
+import org.jetbrains.kotlin.types.AbstractTypeChecker
+
+// See .../types/CastDiagnosticsUtil.kt for counterparts, including isRefinementUseless, isExactTypeCast, isUpcast.
+object FirUselessTypeOperationCallChecker : FirTypeOperatorCallChecker() {
+    override fun check(expression: FirTypeOperatorCall, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (expression.operation !in FirOperation.TYPES) return
+        val arg = expression.argument
+
+        val candidateType = arg.typeRef.coneType.upperBoundIfFlexible().fullyExpandedType(context.session)
+        if (candidateType is ConeKotlinErrorType) return
+
+        val targetType = expression.conversionTypeRef.coneType.fullyExpandedType(context.session)
+        if (targetType is ConeKotlinErrorType) return
+
+        // x as? Type <=> x as Type?
+        val refinedTargetType =
+            if (expression.operation == FirOperation.SAFE_AS) targetType.withNullability(ConeNullability.NULLABLE) else targetType
+        val shouldCheckForExactType = when (expression.operation) {
+            FirOperation.IS, FirOperation.NOT_IS -> false
+            // FirOperation.AS, FirOperation.SAFE_AS
+            else -> shouldCheckForExactType(expression, context)
+        }
+        if (isRefinementUseless(context, candidateType, refinedTargetType, shouldCheckForExactType)) {
+            when (expression.operation) {
+                FirOperation.IS, FirOperation.NOT_IS -> {
+                    val negated = expression.operation == FirOperation.NOT_IS
```

Comment:
I should have mentioned this resembles the variable name in FE 1.0. At the same time, it looks awkward, since it's always negated (as its name indicates!). Renamed to `compileTimeResult`.